File: website/functions/src/models/match.ts
import {UserRecord} from 'firebase-admin/auth';
import {MatchStatus} from '../types/match-status';
import {getDatabase} from 'firebase-admin/database';
import {Player, Players} from './player';
import {Round} from './round';
import {Card} from './card';
import {Hand, Hands} from './hand';

export class Match {
  constructor(
    public id: string,
    public numberOfPlayers: number,
    public maxPoints: number,
    public createdAt: Date,
    public creator: string,
    public status: MatchStatus,
    public players: Players,
    public roundCount: number,
    public round: Round,
  ) {}

  public get playersJoined(): number {
    return this.players.length;
  }

  static new(params: {
    creator: UserRecord,
    numberOfPlayers: number,
    maxPoints: number,
  }): Match {
    const players = new Players();
    players.add(Player.fromUser(params.creator));

    return new Match(
        '',
        params.numberOfPlayers,
        params.maxPoints,
        new Date(),
        params.creator.uid,
        'waitingForPlayers',
        players,
        0,
        new Round(),
    );
  }

  static async load(matchId: string) {
    const snapshot = await getDatabase().ref(`matches/${matchId}`).get();

    if (snapshot.exists()) {
      return Match.parse(snapshot.toJSON());
    } else {
      throw new Error(`Match with ID "${matchId}" does not exist`);
    }
  }

  static parse(data: any): Match {
    return new Match(
        data['id'],
        data['numberOfPlayers'],
        data['maxPoints'],
        new Date(data['createdAt']),
        data['creator'],
        data['status'],
        Players.parse(data['players']),
        data['roundCount'],
        Round.parse(data['round']),
    );
  }

  public async create() {
    this.id = getDatabase().ref('matches').push().key ?? '';
    await this.update();

    return this.id;
  }

  public async join(user: UserRecord) {
    if (this.players.has(user.uid)) {
      return;
    }

    if (this.playersJoined < this.numberOfPlayers) {
      this.players.add(Player.fromUser(user));

      if (this.playersJoined === this.numberOfPlayers) {
        this.status = 'playing';

        for (const player of this.players.list) {
          player.setStatus('playing');
        }

        this.roundCount++;
        this.round = Round.new(this.players);
      }

      await this.update();
    } else {
      throw new Error('Match is full');
    }
  }

  private async update() {
    const matchesRef = getDatabase().ref(`matches/${this.id}`);
    await matchesRef.update(this.json());
  }

  public get isBlocked() {
    return ((this.round.discardPile.length > 0) && this.allPlayersBlocked(this.round.playersHand, this.round.lastCard));
  }

  private allPlayersBlocked(hands: Hands, topCard: Card): boolean {
    for (const hand of hands.list) {
      if (!this.playerBlocked(hand, topCard)) {
        return false;
      }
    }

    return hands.length > 0;
  }

  private playerBlocked(hand: Hand, topCard: Card): boolean {
    if (hand.hiddenPile.length === 0) {
      return !this.canPlayCards(hand.revealedPile, topCard);
    } else {
      return false;
    }
  }

  private canPlayCards(cards: Card[], topCard: Card): boolean {
    for (const card of cards) {
      if (topCard.canAccept(card)) {
        return true;
      }
    }

    return false;
  }

  public removeCard(playerId: string, cardId: string) {
    const hand = this.round.playersHand.of(playerId);
    hand.removeCard(cardId);
  }

  public async playCard(playerId: string, cardId: string) {
    const topCard = this.round.lastCard;
    const revealedPile = this.round.playersHand.of(playerId).revealedPile;

    for (const card of revealedPile) {
      if (card.id === cardId) {
        if (topCard.canAccept(card)) {
          const matchesRef = getDatabase().ref(`matches/${this.id}`);
          await matchesRef.transaction((data: any) => {
            if (data) {
              const match = Match.parse(data);
              match.round.discardPile.push(card);
              match.removeCard(playerId, cardId);

              return match.json();
            } else {
              return this.json();
            }
          }, (error, committed) => {
            if (error) {
              console.log('Transaction failed abnormally!', error);
            } else if (!committed) {
              console.log('We aborted the transaction (because ada already exists).');
            } else {
              console.log('User ada added!');
            }
          },
          true);
          break;
        }
      }
    }
  }

  public json() {
    return {
      id: this.id,
      numberOfPlayers: this.numberOfPlayers,
      maxPoints: this.maxPoints,
      createdAt: this.createdAt,
      creator: this.creator,
      status: this.status,
      players: this.players.json(),
      roundCount: this.roundCount,
      round: this.round.json(),
    };
  }
}

File: server/lib/models/match.dart
import 'package:undervoltage/models/card.dart';
import 'package:undervoltage/models/hand.dart';
import 'package:undervoltage/models/player.dart';
import 'package:undervoltage/models/round.dart';
import 'package:undervoltage/rooms/room.dart';
import 'package:undervoltage/types/match_status.dart';

class Match {
  final String id;
  final Room room;
  final Map<String, Player> players;
  final int roundCount;
  final Round round;
  MatchStatus status;

  Match._({
    required this.id,
    required this.room,
    required this.players,
    required this.roundCount,
    required this.round,
    required this.status,
  });

  factory Match.create(Room room) {
    final Map<String, Player> players = {};

    for (final String playerId in room.players.keys) {
      players[playerId] = Player.create(playerId, room.players[playerId]!);
    }

    return Match._(
      id: room.id,
      room: room,
      status: MatchStatus.playing,
      players: players,
      roundCount: 1,
      round: Round.create(players.values.toList()),
    );
  }

  bool get playerFinished {
    for (final Hand hand in round.playersHand.values) {
      if (hand.finished) {
        return true;
      }
    }

    return false;
  }

  bool get isBlocked =>
      (round.discardPile.isNotEmpty) &&
      allPlayersBlocked(
        round.playersHand.values.toList(),
        round.lastCard,
      );

  bool allPlayersBlocked(List<Hand> hands, Card topCard) {
    for (final Hand hand in hands) {
      if (!playerBlocked(hand, topCard)) {
        return false;
      }
    }

    return hands.isNotEmpty;
  }

  bool playerBlocked(Hand hand, Card topCard) {
    if (hand.hiddenPile.isEmpty) {
      return !canPlayCards(hand.revealedPile, topCard);
    } else {
      return false;
    }
  }

  bool canPlayCards(List<Card> cards, Card topCard) {
    for (final Card card in cards) {
      if (topCard.canAccept(card)) {
        return true;
      }
    }

    return false;
  }

  void endTurn() {
    status = MatchStatus.summary;

    for (final String playerId in round.playersHand.keys) {
      final Hand hand = round.playersHand[playerId]!;
      final Player player = players[playerId]!;
      player.updatePoints(hand);
    }
  }

  void unblock() => round.unblock();

  void removeCard(String playerId, String cardId) {
    final Hand hand = round.playersHand[playerId]!;
    hand.removeCard(cardId);
  }

  void playCard(String playerId, String cardId) {
    final Card topCard = round.lastCard;
    final Hand hand = round.playersHand[playerId]!;

    for (final Card card in hand.revealedPile) {
      if (card.id == cardId) {
        if (topCard.canAccept(card)) {
          round.discardPile.add(card);
          removeCard(playerId, cardId);
        }
      }
    }
  }

  void checkStatus() {
    if (status == MatchStatus.playing) {
      if (playerFinished) {
        endTurn();
      } else if (isBlocked) {
        int limit = round.discardPile.length;

        while (isBlocked && (limit > 1)) {
          unblock();
          limit--;
        }
      }
    }
  }

  void update(double dt) {
    /*if (timeSpent.toInt() % 5 == 0) {
      for (final WebSocket socket in playerIdToWebSocket.values) {
        socket.send(JsonMessage(type: timeSpent.toString()));
      }
    }*/
  }
}
